src/libraries/System.Private.CoreLib/src/System/IO/FileStreamHelpers.Windows.cs

```diff
@@ -1,20 +1,59 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Buffers;
 using System.Diagnostics;
+using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.Win32.SafeHandles;
 
 namespace System.IO
 {
     // this type defines a set of stateless FileStream/FileStreamStrategy helper methods
     internal static class FileStreamHelpers
     {
-        internal static FileStreamStrategy ChooseStrategy(FileStream fileStream, SafeFileHandle handle, FileAccess access, int bufferSize, bool isAsync)
-            => new LegacyFileStreamStrategy(fileStream, handle, access, bufferSize, isAsync);
+        internal const int ERROR_BROKEN_PIPE = 109;
+        internal const int ERROR_NO_DATA = 232;
+        private const int ERROR_HANDLE_EOF = 38;
+        private const int ERROR_IO_PENDING = 997;
 
-        internal static FileStreamStrategy ChooseStrategy(FileStream fileStream, string path, FileMode mode, FileAccess access, FileShare share, int bufferSize, FileOptions options)
-            => new LegacyFileStreamStrategy(fileStream, path, mode, access, share, bufferSize, options);
+        // It's enabled by default. We are going to change that (by removing !) once we fix #16354, #25905 and #24847.
+        internal static bool UseLegacyStrategy { get; }
+            = !AppContextConfigHelper.GetBooleanConfig("System.IO.UseLegacyFileStream", "DOTNET_SYSTEM_IO_USELEGACYFILESTREAM");
```

Comment:
My reasoning behind the negation is that I don't want to have two names:

* one for enabling the new strategy until it's not the default
* one for enabling the old strategy after it's not the default

Here I am just hoping that such a temporary solution is fine. When the switch will happen we are just going to remove the negation. The rename of unit tests projects (called LegacyTests now but testing new implementation in reality) won't also be needed.

@carlossanlop what do you think?

Reply:
I'm not sure I follow. Why are you worried we would need two names? Wouldn't a single appcontext switch suffice for both cases?

Is this being negated because by default we don't expect the user to have this flag explicitly set in their machine, and the default value is zero?

Then what will happen when the new strategy is turned on globally for everyone? Wouldn't they automatically get this appcontext switch set to 1 in their machines after they install the .NET version where we officially enable this feature?